syntax = "proto3";

package snake;

option go_package = "github.com/brensch/snek2/gen/go";

message Point {
  int32 x = 1;
  int32 y = 2;
}

message Snake {
  string id = 1;
  int32 health = 2;
  repeated Point body = 3;
}

message GameState {
  int32 width = 1;
  int32 height = 2;
  repeated Snake snakes = 3;
  repeated Point food = 4;
  string you_id = 5; // ID of the snake to infer for
  int32 turn = 6;
}

message InferenceRequest {
  repeated GameState states = 1;
}

// Represents the prediction for a single GameState
message InferenceResponse {
  // Probabilities for Up, Down, Left, Right (in that order)
  repeated float policy = 1;
  // Win probability (-1 to 1 or 0 to 1)
  float value = 2;
}

message BatchInferenceResponse {
  repeated InferenceResponse responses = 1;
}

service InferenceService {
  rpc Predict(InferenceRequest) returns (BatchInferenceResponse);
}
